Enhancements to Secret Management

maxjeblick opened the pull request:
This pull request adds different backend handlers for storing passwords with `keyring` being the default.
On systems where keyring fails, it is possible to select a local .env backend, or not to save the tokens at all.

**Purpose**
The purpose of these changes is to provide the application with greater flexibility and security when it comes to managing and storing sensitive information.

**Backward compatibility**
A migration process is implemented that converts the old pickle format.
On systems where migration to keyring fails, the user needs to manually reenter missing tokens (this is mentioned in the logs).


Fixes #363

fatihozturkh2o commented on app_utils/utils/setting_utils.py:
qq, what is the difference between `app_utils > sections > settings.py` vs `app_utils > utils > setting_utils.py` ? I wonder if they can be merged then you can revert `from app_utils.utils.utils import X `. utils.utils.utils looks a bit weird 😅

maxjeblick replied:
- `setting_utils.py` contains the backend logic associated with saving and loading the user settings/secrets.
- `settings.py` contains the rendering/routing of the settings page itself.
- `utils.py` contains utils for everything else frontend related.

At some point, `utils.py` should probably be split up into multiple files, as it was done for `llm_studio` utils.
I'd keep the split into `utils` and `setting_utils` as is to start the refactoring of `utils.py`.
Regarding the naming:
`app_utils.utils.utils` should probably be `wave_app.utils.generic_utils` or similar, that is, `app_utils` as a folder name is also a bit misleading.

fatihozturkh2o replied:
I see. I think we can do the following steps to make it more tidy:
- Move the utils folder out of app_utils. (Have them both at the same level)
- Rename app_utils -> wave_utils 
- Then under the utils folder rename setting_utils.py -> settings.py
- Then under the utils folder rename utils.py -> general.py

Doesn't look complicated. Wdyt doing this in another PR?

maxjeblick replied:
Yeah, I agree that renaming/moving components makes sense, in particular, for the frontend/wave part.
Let's probably address this in another PR.

fatihozturkh2o commented:
when I run your PR for the first time, I've seen the followings in the log. Are they all expected?

```
2023/08/16 03:36:57 # {"addr":"127.0.0.1:45378","route":"/18471b8c-b6da-440a-9994-dfb9b2855394","t":"ui_add"}
INFO:     127.0.0.1:57140 - "POST / HTTP/1.1" 200 OK
2023-08-16 03:36:57,827 - INFO: Initializing app ...
2023-08-16 03:36:57,846 - INFO: Initializing app ... done
2023-08-16 03:36:57,846 - INFO: Initializing client None
2023-08-16 03:36:57,854 - INFO: User name: anon
2023-08-16 03:36:57,861 - INFO: Migrating token using Keyring
2023-08-16 03:36:58,099 - INFO: Could not migrate tokens. Please delete /home/fatih/h2o-llmstudio/data/dbs/anon.settings and set your credentials again.Error: 

 Failed to create the collection: Prompt dismissed.. Traceback (most recent call last):
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/secretstorage/util.py", line 48, in send_and_get_reply
    raise DBusErrorResponse(resp_msg)
jeepney.wrappers.DBusErrorResponse: [org.freedesktop.DBus.Error.UnknownMethod] ("No such interface 'org.freedesktop.DBus.Properties' on object at path /org/freedesktop/secrets/collection/login",)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/secretstorage/collection.py", line 177, in get_default_collection
    return Collection(connection)
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/secretstorage/collection.py", line 45, in __init__
    self._collection.get_property('Label')
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/secretstorage/util.py", line 67, in get_property
    (signature, value), = self.send_and_get_reply(msg)
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/secretstorage/util.py", line 52, in send_and_get_reply
    raise ItemNotFoundException('Item does not exist!') from resp
secretstorage.exceptions.ItemNotFoundException: Item does not exist!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/keyring/backends/SecretService.py", line 61, in get_preferred_collection
    collection = secretstorage.get_default_collection(bus)
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/secretstorage/collection.py", line 179, in get_default_collection
    return create_collection(connection, 'Default', 'default', session)
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/secretstorage/collection.py", line 159, in create_collection
    raise PromptDismissedException('Prompt dismissed.')
secretstorage.exceptions.PromptDismissedException: Prompt dismissed.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/fatih/h2o-llmstudio/./app_utils/utils/setting_utils.py", line 278, in _maybe_migrate_to_yaml
    secrets_handler.save(key, user_settings[key])
  File "/home/fatih/h2o-llmstudio/./app_utils/utils/setting_utils.py", line 94, in save
    keyring.set_password(self.namespace, name, password)
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/keyring/core.py", line 61, in set_password
    get_keyring().set_password(service_name, username, password)
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/keyring/backends/SecretService.py", line 87, in set_password
    collection = self.get_preferred_collection()
  File "/home/fatih/h2o-llmstudio/venv/lib/python3.10/site-packages/keyring/backends/SecretService.py", line 63, in get_preferred_collection
    raise InitError("Failed to create the collection: %s." % e)
keyring.errors.InitError: Failed to create the collection: Prompt dismissed..

2023-08-16 03:36:58,099 - INFO: Reading user settings
2023-08-16 03:36:58,138 - ERROR: Could not load password default_aws_access_key from Keyring
2023-08-16 03:36:58,174 - ERROR: Could not load password default_aws_secret_key from Keyring
2023-08-16 03:36:58,208 - ERROR: Could not load password default_kaggle_secret_key from Keyring
2023-08-16 03:36:58,242 - ERROR: Could not load password default_neptune_api_token from Keyring
2023-08-16 03:36:58,275 - ERROR: Could not load password default_huggingface_api_token from Keyring
2023-08-16 03:36:58,307 - ERROR: Could not load password default_openai_api_token from Keyring
2023-08-16 03:36:58,444 - INFO: {'init_app'}
2023-08-16 03:36:58,477 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,515 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,539 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,588 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,718 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,729 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,742 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,760 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,790 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,818 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
2023-08-16 03:36:58,870 - WARNING: Keys {'use_rlhf', 'ppo_clip_value', 'ppo_batch_size', 'adaptive_kl_control', 'ppo_generate_temperature', 'reward_model', 'initial_kl_coefficient', 'kl_horizon', 'ppo_clip_policy', 'kl_target', 'advantages_gamma', 'offload_reward_model', 'scaling_factor_value_loss', 'ppo_epochs', 'advantages_lambda'} are not in the config.
```

maxjeblick commented:
`2023-08-16 03:36:58,099 - INFO: Could not migrate tokens. Please delete /home/fatih/h2o-llmstudio/data/dbs/anon.settings and set your credentials again.Error: `

This warning is expected, see the code for migration. It basically means that keyring storage fails (as it is the case on our remote machines, I wasn't able to configure keyring). The old settings are not deleted in this case.

I'm also happy for suggestions how to better handle this, e.g. using a pop-up that informs the user.


The other warnings `... are not in the config` are unrelated to the PR.

fatihozturkh2o commented:
> `2023-08-16 03:36:58,099 - INFO: Could not migrate tokens. Please delete /home/fatih/h2o-llmstudio/data/dbs/anon.settings and set your credentials again.Error: `
> 
> This warning is expected, see the code for migration. It basically means that keyring storage fails (as it is the case on our remote machines, I wasn't able to configure keyring). The old settings are not deleted in this case.
> 
> I'm also happy for suggestions how to better handle this, e.g. using a pop-up that informs the user.
> 
> The other warnings `... are not in the config` are unrelated to the PR.

I see. So the warnings are from the main brancg?

fatihozturkh2o commented:
> > I see. So the warnings are from the main branch?
> 
> Yes, they arise when calling `load_config_yaml`. Probably better to make the warning more explicit.

they seem to be spamming the logs though. Here I just clicked Home while running my experiment:

![image](https://github.com/h2oai/h2o-llmstudio/assets/70659545/8e29aac9-2283-4fa5-9993-7a3b17ac8d78)

maxjeblick commented:
Created an issue for tracking: https://github.com/h2oai/h2o-llmstudio/issues/372

fatihozturkh2o reviewed (approved):
lgtm!

(you might consider adding type annotation for what the following two functions return though: _get_secrets_handler, _get_usersettings_path)

pascal-pfeiffer reviewed (approved):
Thank you for this security related upgrade, @maxjeblick 

I still ran into issues when just upgrading to the PR branch, but using a clean environment everything was working fine. I think we can live with that. 

lgtm

maxjeblick commented:
Sorry for the late reply @pascal-pfeiffer, I ran into some new timeout issues after I tried to configure keyring on the remote machine. Compared to the reviewed PR, I added the following:

- Default `credential_saver` was changed to .env file to avoid any keyring issues. This should allow for migration without issues.
- Keyring handler will be disabled if it is not configured correctly (`keyring.get_password` fails for some reason). While this adds up to 3 seconds time overhead when starting the app, it will safeguard against potential hang-ups.
- I fixed merge conflicts with latest main

Please feel free to re-review the changes.
